cpp/src/arrow/dataset/discovery.cc

```diff
@@ -217,7 +217,7 @@ Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
   // ARROW-12481.
   std::string internal_path;
   ARROW_ASSIGN_OR_RAISE(std::shared_ptr<fs::FileSystem> filesystem,
-                        fs::FileSystemFromUri(uri, &internal_path))
+                        fs::FileSystemFromUriAndFs(uri, &internal_path, options.file_system_java))
```

Comment:
I think what you would do is, if a filesystem is provided, call FileSystem::PathFromUri to set `internal_path`. If it succeeds, then proceed, else raise an error.